### PR TITLE
rgw: added zipper tracer driver

### PR DESCRIPTION
**Pull request by dpaganel**
rgw Tracer Filter: The Tracer Filter is an in-progress project to enable a pass-through filter on the rgw sal abstractions. Activate with -WITH_RADOSGW_TRACER flag during cmake. Introduction of the filter includes modification to underlying classes User and Bucket to have exclusively abstract functions. Rebased version of earlier pull request for new main branch.

Signed-off-by: Daniel Paganelli <dpaganel@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by alimaredia** [comment]
jenkins test api

**Comment by dpaganel** [comment]
jenkins test api

**Comment by dpaganel** [comment]
jenkins test api

**Comment by alimaredia** [comment]
jenkins test api

**Comment by dpaganel** [comment]
jenkins test api

File: src/rgw/rgw_sal.h [hunk 1]
```diff
@@ -43,7 +46,7 @@ namespace rgw {
   namespace IAM { struct Policy; }
 }
 
-class RGWGetDataCB {
+class RGWGetDataCB {\
```
Comment (cbodley) [inline src/rgw/rgw_sal.h:49]: typo

File: src/rgw/rgw_sal.h [hunk 2]
```diff
@@ -476,6 +479,13 @@ class User {
     User(User& _o) = default;
     virtual ~User() = default;
 
+
+    /*returns whether info is empty, as part of making Bucket fully abstract -Daniel P*/
+    virtual bool info_empty()
+    {
+      return info.user_id.id.empty();
```
Comment (cbodley) [inline src/rgw/rgw_sal.h:486]: ok, so the problem with `if (rgw::sal::User::empty(user))` is that it's not polymorphic:
```c++
static bool empty(User* u) { return (!u || u->info.user_id.id.empty()); }
static bool empty(std::unique_ptr<User>& u) { return (!u || u->info.user_id.id.empty()); }
```
can these `empty()` functions just use your virtual `info_empty()` function instead?
```c++
static bool empty(User* u) { return (!u || u->info_empty()); }
static bool empty(std::unique_ptr<User>& u) { return (!u || u->info_empty()); }
```
that way you don't have to change every caller of `rgw::sal::User::empty()`

p.s. for @dang, is there really a use-case for a default-constructed `User`? can't we just make it a precondition, that you don't construct a `User` until you have a uid for it? that way we don't have this extra uninitialized state, and don't need `info_empty()` at all

File: src/rgw/rgw_sal.h [hunk 3]
```diff
@@ -504,23 +514,23 @@ class User {
     /** Get the display name for this User */
     virtual std::string& get_display_name() { return info.display_name; }
     /** Get the tenant name for this User */
-    const std::string& get_tenant() { return info.user_id.tenant; }
+    virtual const std::string& get_tenant() { return info.user_id.tenant; }
```
Comment (cbodley) [inline src/rgw/rgw_sal.h:517]: @dang now that we can have filter stores that wrap other stores, isn't it time to get rid of the member variables in the sal base classes?

if we have a `TracerUser` wrapping a `RadosUser`, we shouldn't store a `RGWUserInfo info` in both - that's bound to cause confusion and bugs

now that these getter functions are virtual, we can make them pure virtual in `rgw::sal::User` and move the member variables into the backends like `RadosUser`/`DBUser`/`MotrUser`. same with the other sal base classes

Comment (mattbenjamin) [reply src/rgw/rgw_sal.h:517]: agree, have already requested this

File: src/rgw/rgw_sal.h (hunk 2 above)
Comment (mattbenjamin) [reply src/rgw/rgw_sal.h:486]: good idea

File: src/rgw/rgw_sal.h (hunk 3 above)
Comment (dang) [reply src/rgw/rgw_sal.h:517]: Already working on this, mostly complete.  Was waiting for this PR, so as not to break all Daniel's work.

**Comment by idryomov** [comment]
@dpaganel This PR now has a lot of commits that have nothing to do with rgw.  Please drop the "Merge branch 'main' into wip-dpaganel-tracer" commit and rebase just your rgw changes on top of the `main` branch.

**Pull request closed without merging**